automerge-persistent-fs/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+
+#[derive(Debug)]
+pub struct FsPersister {
+    changes_path: PathBuf,
+    doc_path: PathBuf,
+    sync_states_path: PathBuf,
+    cache: FsPersisterCache,
+    sizes: StoredSizes,
+}
+
+#[derive(Debug)]
+pub struct FsPersisterCache {
+    changes: HashMap<(ActorId, u64), Vec<u8>>,
+    document: Option<Vec<u8>>,
+    sync_states: HashMap<Vec<u8>, Vec<u8>>,
+}
+
+impl FsPersisterCache {
+    fn flush_changes_sync(&mut self, changes_path: PathBuf) -> Result<usize, std::io::Error> {
+        let mut flushed = 0;
+        for ((a, s), c) in self.changes.drain() {
+            fs::write(make_changes_path(&changes_path, &a, s), &c)?;
+            flushed += c.len();
+        }
+        Ok(flushed)
+    }
+
+    fn flush_document_sync(&mut self, doc_path: PathBuf) -> Result<usize, std::io::Error> {
+        let mut flushed = 0;
+        if let Some(data) = self.document.take() {
+            fs::write(&doc_path, &data)?;
+            flushed = data.len();
+        }
+        Ok(flushed)
+    }
+
+    fn flush_sync_states_sync(
+        &mut self,
+        sync_states_path: PathBuf,
+    ) -> Result<usize, std::io::Error> {
+        let mut flushed = 0;
+        for (peer_id, sync_state) in self.sync_states.drain() {
+            fs::write(make_peer_path(&sync_states_path, &peer_id), &sync_state)?;
+            flushed += sync_state.len();
+        }
+        Ok(flushed)
+    }
+
+    #[cfg(feature = "async")]
+    async fn flush_changes(&mut self, changes_path: PathBuf) -> Result<usize, std::io::Error> {
+        let futs = futures::stream::FuturesUnordered::new();
+        for ((a, s), c) in self.changes.drain() {
+            let len = c.len();
+            futs.push(
+                tokio::fs::write(make_changes_path(&changes_path, &a, s), c).map(move |_| Ok(len)),
+            );
+        }
+        let res: Result<Vec<usize>, std::io::Error> = futs.try_collect().await;
+        Ok(res?.iter().sum())
+    }
+
+    #[cfg(feature = "async")]
+    async fn flush_document(&mut self, doc_path: PathBuf) -> Result<usize, std::io::Error> {
+        let mut flushed = 0;
+        if let Some(data) = self.document.take() {
+            tokio::fs::write(&doc_path, &data).await?;
+            flushed = data.len();
+        }
+        Ok(flushed)
+    }
+
+    #[cfg(feature = "async")]
+    async fn flush_sync_states(
+        &mut self,
+        sync_states_path: PathBuf,
+    ) -> Result<usize, std::io::Error> {
+        let futs = futures::stream::FuturesUnordered::new();
+        for (peer_id, sync_state) in self.sync_states.drain() {
+            let len = sync_state.len();
+            futs.push(
+                tokio::fs::write(make_peer_path(&sync_states_path, &peer_id), sync_state)
+                    .map(move |_| Ok(len)),
+            );
+        }
+        let res: Result<Vec<usize>, std::io::Error> = futs.try_collect().await;
+        Ok(res?.iter().sum())
+    }
+
+    #[cfg(feature = "async")]
+    pub async fn flush(
```

Comment:
Could we switch the `_sync` and async variants? Seems strange to have the non-default have an extension on the name.